impeller/entity/contents/atlas_contents.cc

```diff
@@ -0,0 +1,145 @@
+// found in the LICENSE file.
+
+#include <iostream>
+#include <optional>
+#include "impeller/geometry/path_builder.h"
+#include "impeller/renderer/formats.h"
+#include "impeller/renderer/sampler_library.h"
+#include "impeller/renderer/vertex_buffer_builder.h"
+#include "linear_gradient_contents.h"
+
+#include "impeller/entity/atlas_fill.frag.h"
+#include "impeller/entity/atlas_fill.vert.h"
+#include "impeller/entity/contents/atlas_contents.h"
+#include "impeller/entity/contents/content_context.h"
+#include "impeller/entity/contents/solid_color_contents.h"
+#include "impeller/entity/entity.h"
+#include "impeller/renderer/render_pass.h"
+
+namespace impeller {
+
+AtlasContents::AtlasContents() = default;
+
+AtlasContents::~AtlasContents() = default;
+
+void AtlasContents::SetTexture(std::shared_ptr<Texture> texture) {
+  texture_ = std::move(texture);
+}
+
+std::shared_ptr<Texture> AtlasContents::GetTexture() const {
+  return texture_;
+}
+
+void AtlasContents::SetXForm(std::vector<Matrix> xform) {
+  xform_ = xform;
+}
+
+void AtlasContents::SetTextureCoordinates(std::vector<Rect> texture_coords) {
+  texture_coords_ = texture_coords;
+}
+
+void AtlasContents::SetColors(std::vector<Color> colors) {
+  colors_ = colors;
+}
+
+void AtlasContents::SetBlendMode(Entity::BlendMode blend_mode) {
+  // TODO(jonahwilliams): blending of colors with texture.
+  blend_mode_ = blend_mode;
+}
+
+std::optional<Rect> AtlasContents::GetCoverage(const Entity& entity) const {
+  // TODO(jonahwilliams): compute coverage of atlas.
+  return std::nullopt;
+}
+
+void AtlasContents::SetSamplerDescriptor(SamplerDescriptor desc) {
+  sampler_descriptor_ = std::move(desc);
+}
+
+const SamplerDescriptor& AtlasContents::GetSamplerDescriptor() const {
+  return sampler_descriptor_;
+}
+
+bool AtlasContents::Render(const ContentContext& renderer,
+                           const Entity& entity,
+                           RenderPass& pass) const {
+  if (texture_ == nullptr) {
+    return true;
+  }
+
+  using VS = AtlasFillVertexShader;
+  using FS = AtlasFillFragmentShader;
+
+  const auto coverage_rect = path_.GetBoundingBox();
+
+  if (!coverage_rect.has_value()) {
+    return true;
+  }
+
+  if (coverage_rect->size.IsEmpty()) {
+    return true;
+  }
+
+  const auto texture_size = texture_->GetSize();
+  if (texture_size.IsEmpty()) {
+    return true;
+  }
+
+  VertexBufferBuilder<VS::PerVertexData> vertex_builder;
+  {
+    vertex_builder.Reserve(texture_coords_.size() * 6);
+    constexpr size_t indexes[6] = {0, 1, 2, 1, 2, 3};
+    constexpr Scalar width[6] = {0, 1, 0, 1, 0, 1};
+    constexpr Scalar height[6] = {0, 0, 1, 0, 1, 1};
+    for (size_t i = 0; i < texture_coords_.size(); i++) {
+      auto sample_rect = texture_coords_[i];
+      auto matrix = xform_[i];
+      auto color = (colors_.size() > 0 ? colors_[i] : Color::Black());
+      auto transformed_points = sample_rect.GetTransformedPoints(matrix);
```

Comment:
Since `transformed_points` is used as the position of the draw rect, should this be something like `Rect::MakeSize(sample_rect.size).GetTransformedPoints(matrix)`? It seems like the transformed rect should have origin 0 rather than the position used for sampling the atlas.

Reply:
Ahh this was it, thanks @bdero !